src/geometamaker/models.py

```diff
@@ -0,0 +1,584 @@
+    rows: int
+    # without post-init, schema ends up as a dict, or whatever is passed in.
+    schema: TableSchema = dataclasses.field(default_factory=TableSchema)
+
+    def __post_init__(self):
+        super().__post_init__()
+        # Allow init of the resource with a schema of type
+        # TableSchema, or type dict. Mostly because dataclasses.replace
+        # calls init, but the base object will have already been initialized.
+        if isinstance(self.schema, TableSchema):
+            return
+        self.schema = TableSchema(**self.schema)
+
+    def _get_field(self, name):
+        """Get an attribute by its name property.
+
+        Args:
+            name (string): to match the value of the 'name' key in a dict
+
+        Returns:
+            tuple of (list index of the matching attribute, the attribute
+                dict)
+
+        Raises:
+            KeyError if no attributes exist in the resource or if the named
+                attribute does not exist.
+
+        """
+        if len(self.schema.fields) == 0:
+            raise KeyError(
+                f'{self.schema} has no fields')
+        for idx, field in enumerate(self.schema.fields):
+            if field.name == name:
+                return idx, field
+        raise KeyError(
+            f'{self.schema} has no field named {name}')
+
+    def set_field_description(self, name, title=None, description=None,
+                              units=None, type=None):
+        """Define metadata for a tabular field.
+
+        Args:
+            name (str): name and unique identifier of the field
+            title (str): title for the field
+            description (str): description of the field
+            units (str): unit of measurement for the field's values
+            type (str): datatype of values in the field
+
+        """
+        idx, field = self._get_field(name)
+
+        if title is not None:
+            field.title = title
+        if description is not None:
+            field.description = description
+        if units is not None:
+            field.units = units
+        if type is not None:
+            field.type = type
+
+        self.schema.fields[idx] = field
+
+    def get_field_description(self, name):
+        """Get the attribute metadata for a field.
+
+        Args:
+            name (str): name and unique identifier of the field
+
+        Returns:
+            FieldSchema
+        """
+        idx, field = self._get_field(name)
+        return field
+
+
+@dataclass(kw_only=True)
+class ArchiveResource(Resource):
+    """Class for metadata for an archive resource."""
+
+    compression: str
+    innerpath: str
+
+
+@dataclass(kw_only=True)
+class VectorResource(TableResource):
+    """Class for metadata for a vector resource."""
+
+    spatial: SpatialSchema
+
+
+@dataclass(kw_only=True)
+class RasterResource(Resource):
+    """Class for metadata for a raster resource."""
+
+    schema: RasterSchema
+    spatial: SpatialSchema
```

Comment:
The inheritance hierarchy for rasters and vectors makes sense to me given the database programming model of a vector.  Raster Attribute Tables might be another interesting sort of challenge, though ... should we handle RATs in some way?  Not necessarily advocating for handling a RAT in a different way than the normal table would be handled, just wanted to ask!